src/jsTypeFactory.cc

```diff
@@ -314,52 +314,110 @@ void setPyException(JSContext *cx) {
     JS::RootedValue jsExceptionValue(cx, JS::ObjectValue(*jsException));
     JS_SetPendingException(cx, jsExceptionValue);
   }
+  return true;
 }
 
 bool callPyFunc(JSContext *cx, unsigned int argc, JS::Value *vp) {
   JS::CallArgs callargs = JS::CallArgsFromVp(argc, vp);
 
   // get the python function from the 0th reserved slot
-  JS::Value pyFuncVal = js::GetFunctionNativeReserved(&(callargs.callee()), 0);
-  PyObject *pyFunc = (PyObject *)(pyFuncVal.toPrivate());
-
-  unsigned int callArgsLength = callargs.length();
+  PyObject *pyFunc = (PyObject *)js::GetFunctionNativeReserved(&(callargs.callee()), 0).toPrivate();
+  Py_INCREF(pyFunc);
+  PyObject *pyRval = NULL;
+  PyObject *pyArgs = NULL;
+  Py_ssize_t nNormalArgs = 0;   // number of positional non-default arguments
+  Py_ssize_t nDefaultArgs = 0;  // number of positional default arguments
+  bool varargs = false;
+  bool unknownNargs = false;
+
+  if (PyCFunction_Check(pyFunc)) {
+    if (((PyCFunctionObject *)pyFunc)->m_ml->ml_flags & METH_NOARGS) { // 0 arguments
```

Comment:
can we evaluate this complex expression only once and cache it aside in a local?